### PR TITLE
Initial commit of kext extraction code

### PR DESCRIPTION
**Pull request by aixiansheng**

### CODE REVIEW

**Comment by blacktop** [comment]
Why are you putting all the kext stuff in the kernelcache file and not the kext file?

**Comment by aixiansheng** [comment]
Cuz the Prelink info is kernelcache-wide.  All of the kexts point to the same blob.

**Comment by blacktop** [comment]
I have refactored pkg/macho into blacktop/go-macho now

File: pkg/kernelcache/kext.go [hunk 1]
```diff
@@ -2,69 +2,220 @@ package kernelcache
...
-
-	"github.com/blacktop/go-macho"
-	"github.com/blacktop/go-plist"
 )
 
-type prelinkInfo struct {
-	PrelinkInfoDictionary []CFBundle `plist:"_PrelinkInfoDictionary,omitempty"`
+// KextList lists all the kernel extensions in the kernelcache
+func KextList(kernel string) error {
+	kc, err := NewKernelCache(kernel)
```
Comment (blacktop) [inline pkg/kernelcache/kext.go:13]: Check failure on line 13 in pkg/kernelcache/kext.go

 
GitHub Actions
/ build (windows-latest)

pkg/kernelcache/kext.go#L13
undefined: NewKernelCache

File: pkg/kernelcache/kext.go [hunk 2]
```diff
@@ -2,69 +2,220 @@ package kernelcache
...
+	}
+	defer kc.Close()
+
+	prelink, err := kc.PrelinkInfoDict()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("FOUND:", len(prelink.PrelinkInfoDictionary))
+	prelink.ForeachBundle(func(bundle * CFBundle) error {
```
Comment (blacktop) [inline pkg/kernelcache/kext.go:25]: Check failure on line 25 in pkg/kernelcache/kext.go

 
GitHub Actions
/ build (windows-latest)

pkg/kernelcache/kext.go#L25
undefined: CFBundle

File: pkg/kernelcache/kext.go [hunk 3]
```diff
@@ -2,69 +2,220 @@ package kernelcache
...
-	defer m.Close()
 
-	for _, sec := range m.Sections {
-		if sec.Seg == "__PRELINK_INFO" && sec.Name == "__info" {
-			f, err := os.Open(kernel)
-			if err != nil {
-				return err
+	defer fh.Close()
+
+	kc, err := NewKernelCache(kernel)
```
Comment (blacktop) [inline pkg/kernelcache/kext.go:92]: Check failure on line 92 in pkg/kernelcache/kext.go

 
GitHub Actions
/ build (windows-latest)

pkg/kernelcache/kext.go#L92
undefined: NewKernelCache

File: pkg/kernelcache/kext.go [hunk 4]
```diff
@@ -49,6 +47,39 @@ func unSeek(fh * os.File, thingy func() error) error {
 	return err
 }
 
+type Segment64 struct {
```
Comment (blacktop) [inline pkg/kernelcache/kext.go:50]: Use the go-macho/types

File: pkg/kernelcache/kext.go [hunk 5]
```diff
@@ -49,6 +47,39 @@ func unSeek(fh * os.File, thingy func() error) error {
...
+	Align     uint32
+	Reloff    uint32
+	Nreloc    uint32
+	Flags     uint32
+	Reserved1 uint32
+	Reserved2 uint32
+	Reserved3 uint32
+}
+
+func roundUp64(x, y uint64) uint64 {
```
Comment (blacktop) [inline pkg/kernelcache/kext.go:79]: same

**Pull request merged**